src/engine/strat_engine/metadata.rs

```diff
@@ -4,262 +4,416 @@
+    }
+
+    // Write data to the older region
+    pub fn save_state(&mut self, time: &Timespec, data: &[u8], f: &mut File) -> EngineResult<()> {
+        let region_size: u64 = *self.region_size * SECTOR_SIZE;
+
+        let mut save_region = |region: usize| -> EngineResult<()> {
+            let hdr_buf = MDAHeader::to_buf(data, time);
+            let offset = BDA_STATIC_HDR_SIZE + (region as u64 * region_size);
+
+            if MDA_REGION_HDR_SIZE + data.len() > region_size as usize {
+                return Err(EngineError::Engine(ErrorEnum::Invalid,
+                                               "data larger than region_size".into()));
+            }
+
+            try!(f.seek(SeekFrom::Start(offset)));
+            try!(f.write_all(&hdr_buf));
+            try!(f.write_all(data));
+            try!(f.flush());
+
+            Ok(())
+        };
+
+        let older_region = self.older();
+
+        if let Some(updated) = self.mdas[older_region].last_updated {
+            if updated >= *time {
+                return Err(EngineError::Engine(ErrorEnum::Invalid,
+                                               "Overwriting newer data".into()));
+            }
         }
+
+        // Save to primary and backup regions
+        // TODO: Should we ignore errors?
+        try!(save_region(older_region));
+        try!(save_region(older_region + 2));
+
+        self.mdas[older_region].last_updated = Some(*time);
+        self.mdas[older_region].used = data.len() as u64;
+
+        Ok(())
     }
 
-    /// Write SigBlock to buf at offset.
-    pub fn write(&self, buf: &mut [u8], offset: usize) {
-        buf[offset + 4..offset + 20].clone_from_slice(STRAT_MAGIC);
-        LittleEndian::write_u64(&mut buf[offset + 20..offset + 28], *self.total_size);
-        buf[offset + 32..offset + 64]
-            .clone_from_slice(self.pool_uuid.simple().to_string().as_bytes());
-        buf[offset + 64..offset + 96]
-            .clone_from_slice(self.dev_uuid.simple().to_string().as_bytes());
-        self.mda.write(buf, offset + 96);
-        LittleEndian::write_u32(&mut buf[offset + 160..offset + 164],
-                                *self.mda_sectors as u32);
-        LittleEndian::write_u32(&mut buf[offset + 164..offset + 168],
-                                *self.reserved_sectors as u32);
-
-        let hdr_crc = crc32::checksum_ieee(&buf[offset + 4..offset + SECTOR_SIZE as usize]);
-        LittleEndian::write_u32(&mut buf[offset..offset + 4], hdr_crc);
+    pub fn load_state(&self, f: &mut File) -> EngineResult<Option<Vec<u8>>> {
+        let newer_region = self.newer();
+        let mda = &self.mdas[newer_region];
+
+        let mut load_json = |region| {
+            let offset = BDA_STATIC_HDR_SIZE + (region as u64 * *self.region_size * SECTOR_SIZE) +
+                         MDA_REGION_HDR_SIZE as u64;
+            try!(f.seek(SeekFrom::Start(offset)));
+            mda.load_region(f)
+        };
+
+        Ok(load_json(newer_region)
```

Comment:
I've opened https://github.com/stratis-storage/stratisd/issues/210, there are a few places we could repair stuff, but it's probably good to think a bit more before committing to this.